File: src/server/controllers/maps.ts
import crypto from 'crypto';
import express from 'express';
import fsExtra from 'fs-extra';
import path from 'path';

import { getCreaturesInTile } from '@/lib/azerothcore-client/creatures';
import { exportTexture } from '@/lib/converter/character/utils';
import { getListFiles, registerListfileClearHook } from '@/lib/wow/listfile-cache';
import { FileEntry, MapListItem, wowDataClient } from '@/lib/wow-data-client/wow-data-client';
import { assertDesktopOnly, desktopOnlyStatus } from '@/server/shared-hosting';
import {
  applyCascBuildCache, etagFromParts, matchNotModified, minimapPngPath, writeNotModified,
} from '@/server/utils/casc-cache';

import { registerMapGenerateRoutes } from './maps-generate';

type TileInfo = { x: number; y: number; hasTexture: boolean };
type MapWithTiles = MapListItem & { tiles: TileInfo[] };

const tileBlpRegex = /^world\/minimaps\/([^/]+)\/map(\d{1,2})_(\d{1,2})\.blp$/i;
const tileAdtRegex = /^world\/maps\/([^/]+)\/\1_(\d{2})_(\d{2})\.adt$/i;

let mapsWithTiles: MapWithTiles[] | null = null;
const mapsByDir = new Map<string, MapWithTiles>(); // dir(lowercased) -> map with tiles
const fileNameToEntry = new Map<string, FileEntry>(); // normalized lowercased path -> entry

registerListfileClearHook(() => {
  mapsWithTiles = null;
  mapsByDir.clear();
  fileNameToEntry.clear();
});

async function ensureMapsIndex(): Promise<void> {
  if (mapsWithTiles !== null) return;
  await buildMapsIndex();
}

async function buildMapsIndex(): Promise<void> {
  await wowDataClient.waitUntilReady();

  let baseMaps: MapListItem[] = [];
  try {
    baseMaps = await wowDataClient.getMapList();
  } catch {
    baseMaps = [];
  }
  const files = await getListFiles();

  const adtByDir = new Map<string, Set<string>>(); // key: "x,y"
  const texByDir = new Map<string, Set<string>>();

  for (const fileEntry of files) {
    const fileName = fileEntry.fileName.replace(/\\/g, '/').toLowerCase();

    // Track minimap textures
    const matchesBlp = tileBlpRegex.exec(fileName);
    if (matchesBlp) {
      const dir = matchesBlp[1].toLowerCase();
      const x = parseInt(matchesBlp[2], 10);
      const y = parseInt(matchesBlp[3], 10);
      if (x >= 0 && x < 64 && y >= 0 && y < 64) {
        let set = texByDir.get(dir);
        if (!set) { set = new Set<string>(); texByDir.set(dir, set); }
        set.add(`${x},${y}`);
        fileNameToEntry.set(fileName, fileEntry);
      }
      continue;
    }

    // Track ADT presence
    const matchesAdt = tileAdtRegex.exec(fileName);
    if (matchesAdt) {
      const dir = matchesAdt[1].toLowerCase();
      const x = parseInt(matchesAdt[2], 10);
      const y = parseInt(matchesAdt[3], 10);
      if (x >= 0 && x < 64 && y >= 0 && y < 64) {
        let set = adtByDir.get(dir);
        if (!set) { set = new Set<string>(); adtByDir.set(dir, set); }
        set.add(`${x},${y}`);
      }
    }
  }

  mapsWithTiles = [];
  mapsByDir.clear();

  for (const m of baseMaps) {
    const dir = m.dir.toLowerCase();
    const tilesMap = new Map<string, TileInfo>();

    const adtSet = adtByDir.get(dir);
    if (adtSet) {
      for (const key of adtSet) {
        const [xs, ys] = key.split(',');
        const x = parseInt(xs, 10);
        const y = parseInt(ys, 10);
        tilesMap.set(key, { x, y, hasTexture: false });
      }
    }

    const texSet = texByDir.get(dir);
    if (texSet) {
      for (const key of texSet) {
        const [xs, ys] = key.split(',');
        const x = parseInt(xs, 10);
        const y = parseInt(ys, 10);
        const prev = tilesMap.get(key);
        if (prev) {
          prev.hasTexture = true;
        } else {
          tilesMap.set(key, { x, y, hasTexture: true });
        }
      }
    }

    const tiles = Array.from(tilesMap.values());
    const withTiles: MapWithTiles = { ...m, tiles };
    mapsWithTiles.push(withTiles);
    mapsByDir.set(dir, withTiles);
  }
  // console.log(`Total maps: ${mapsWithTiles.length}, total map tiles: ${_.sumBy(mapsWithTiles, 'tiles.length')}`);
}

export function ControllerMaps(router: express.Router) {
  void buildMapsIndex();

  registerMapGenerateRoutes(router, (key) => {
    const entry = mapsByDir.get(key);
    return entry ? { id: entry.id, dir: entry.dir } : undefined;
  });

  // GET /api/maps -> list maps
  router.get('/maps', async (req, res) => {
    try {
      await ensureMapsIndex();
      const list = mapsWithTiles ?? [];
      const buildKey = wowDataClient.cascInfo?.buildKey ?? '';
      const etag = etagFromParts('maps', buildKey, String(list.length));
      if (matchNotModified(req, etag)) {
        applyCascBuildCache(res, req, buildKey, etag);
        return writeNotModified(res, etag);
      }
      applyCascBuildCache(res, req, buildKey, etag);
      return res.json(list.map((m) => ({
        id: m.id, name: m.name, dir: m.dir, expansionID: m.expansionID,
      })));
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });

  // GET /api/maps/:map/wdt-mask -> tiles list with hasTexture flags
  router.get('/maps/:map/wdt-mask', async (req, res) => {
    try {
      await ensureMapsIndex();
      const key = String(req.params.map).toLowerCase();
      const entry = mapsByDir.get(key);
      const tiles = entry?.tiles ?? [];
      const buildKey = wowDataClient.cascInfo?.buildKey ?? '';
      const etag = etagFromParts('wdt-mask', buildKey, key, String(tiles.length));
      if (matchNotModified(req, etag)) {
        applyCascBuildCache(res, req, buildKey, etag);
        return writeNotModified(res, etag);
      }
      applyCascBuildCache(res, req, buildKey, etag);
      return res.json({ map: req.params.map, size: 64, tiles });
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });

  // POST /api/maps/:map/creatures-check -> count creatures in selected tiles only
  router.post('/maps/:map/creatures-check', async (req, res) => {
    try {
      assertDesktopOnly();
      const key = String(req.params.map).toLowerCase();
      const entry = mapsByDir.get(key);
      if (!entry) {
        return res.status(404).json({ error: 'Unknown map' });
      }

      const tiles = req.body?.tiles;
      if (!Array.isArray(tiles) || tiles.length === 0) {
        return res.status(400).json({ error: 'Invalid request body' });
      }

      const seen = new Set<string>();
      const checkedTiles: { x: number; y: number }[] = [];
      let creatureCount = 0;
      for (const tile of tiles) {
        const x = Number(tile?.x);
        const y = Number(tile?.y);
        if (!Number.isInteger(x) || !Number.isInteger(y) || x < 0 || x >= 64 || y < 0 || y >= 64) {
          return res.status(400).json({ error: 'Tile coordinates must be within 0..63' });
        }
        const tileKey = `${x},${y}`;
        if (seen.has(tileKey)) continue;
        seen.add(tileKey);
        checkedTiles.push({ x, y });
        const creatures = await getCreaturesInTile(entry.id, [x, y]);
        creatureCount += creatures.length;
      }

      if (checkedTiles.length === 0) {
        return res.status(400).json({ error: 'No valid tiles provided' });
      }

      return res.json({
        hasCreatures: creatureCount > 0,
        creatureCount,
        checkedTiles,
      });
    } catch (e) {
      const err = e as Error;
      const status = desktopOnlyStatus(err);
      if (status === 403) {
        return res.status(403).json({ error: err.message });
      }
      const message = err.message;
      if (message.includes('azerothcore database not found') || message.includes('ENOENT')) {
        return res.json({ hasCreatures: false, creatureCount: 0, checkedTiles: [] });
      }
      return res.status(500).json({ error: message });
    }
  });

  // GET /api/maps/:map/minimap/:x/:y -> PNG bytes
  // The minimap BLP is decoded to PNG via the pipeline-aware exportTexture helper.
  router.get('/maps/:map/minimap/:x/:y', async (req, res) => {
    try {
      const { map } = req.params;
      const x = parseInt(String(req.params.x), 10);
      const y = parseInt(String(req.params.y), 10);

      if (!(x >= 0 && x < 64 && y >= 0 && y < 64)) {
        return res.status(400).json({ error: 'x and y must be within 0..63' });
      }

      await wowDataClient.waitUntilReady();

      // Normalize directory and coordinates to match Blizzard naming
      const mapDir = String(map).toLowerCase();
      const xs = x.toString().padStart(2, '0');
      const ys = y.toString().padStart(2, '0');

      const buildKey = wowDataClient.cascInfo?.buildKey || '';
      const etagSeed = `${buildKey}|${map}|${x}|${y}`;
      const quotedETag = `"${crypto.createHash('md5').update(etagSeed).digest('hex')}"`;

      if (matchNotModified(req, quotedETag)) {
        applyCascBuildCache(res, req, buildKey, quotedETag, true);
        return writeNotModified(res, quotedETag);
      }

      const sendPng = async (pngPath: string) => {
        res.setHeader('Content-Type', 'image/png');
        applyCascBuildCache(res, req, buildKey, quotedETag, true);
        return res.send(await fsExtra.readFile(path.resolve(pngPath)));
      };

      // If PNG already exists in the export asset directory, serve it directly.
      const assetDir = await wowDataClient.getAssetDir();
      const preexistingPng = minimapPngPath(assetDir, buildKey, mapDir, xs, ys);
      if (fsExtra.existsSync(preexistingPng)) {
        return sendPng(preexistingPng);
      }

      // Resolve the BLP using the prebuilt hash table
      const blpPath = `world/minimaps/${mapDir}/map${xs}_${ys}.blp`;
      const file = fileNameToEntry.get(blpPath);
      if (!file?.fileDataID) {
        return res.status(404).json({ error: 'Minimap tile not found' });
      }
      const relPng = await exportTexture(file.fileDataID);
      const pngPath = path.join(assetDir, relPng);
      return sendPng(pngPath);
    } catch (e) {
      return res.status(500).json({ error: e instanceof Error ? e.message : String(e) });
    }
  });
}

File: golang/internal/server/api/maps.go
package api

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/pqhuy98/wow-converter/internal/azerothcore"
	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/converter/runtimecache"
	"github.com/pqhuy98/wow-converter/internal/formats/blp"
	"github.com/pqhuy98/wow-converter/internal/wow/casc"
	"github.com/pqhuy98/wow-converter/internal/wow/client"
)

type tileInfo struct {
	X          int  `json:"x"`
	Y          int  `json:"y"`
	HasTexture bool `json:"hasTexture"`
}

type mapWithTiles struct {
	casc.MapEntry
	Tiles []tileInfo `json:"tiles"`
}

var (
	mapsMu         sync.RWMutex
	mapsIndexMu    sync.Mutex
	mapsWithTiles  []mapWithTiles
	mapsByDir     = map[string]*mapWithTiles{}
	fileNameIndex = map[string]casc.ListfileEntry{}

	tileBlpRegex = regexp.MustCompile(`^world/minimaps/([^/]+)/map(\d{1,2})_(\d{1,2})\.blp$`)
	tileAdtRegex = regexp.MustCompile(`(?i)^world/maps/([^/]+)/([^/]+)_(\d{2})_(\d{2})\.adt$`)
)

func resetMapsFileIndex() {
	mapsIndexMu.Lock()
	defer mapsIndexMu.Unlock()
	mapsMu.Lock()
	defer mapsMu.Unlock()
	mapsWithTiles = nil
	mapsByDir = map[string]*mapWithTiles{}
	fileNameIndex = map[string]casc.ListfileEntry{}
}

func init() {
	runtimecache.RegisterConverterClearHook(resetMapsFileIndex)
}

func buildMapsIndex(ctx context.Context, d *Deps) error {
	baseMaps, err := d.Client.GetMapList(ctx)
	if err != nil {
		baseMaps = nil
	}

	start := time.Now()
	var files []casc.ListfileEntry
	if direct, ok := d.Client.(client.DirectListfileClient); ok {
		log.Printf("Building maps index from listfile...")
		var err error
		files, err = direct.CollectMapTileFileIndex(ctx)
		if err == nil {
			log.Printf("Collected %d map tile files in %.1fs", len(files), time.Since(start).Seconds())
		} else {
			log.Printf("Direct map tile index unavailable (%v), falling back to full listfile copy", err)
			files = nil
		}
	}
	if files == nil {
		log.Printf("Building maps index (full listfile copy)...")
		var err error
		files, err = getListFiles(ctx, d.Client)
		if err != nil {
			return err
		}
		log.Printf("Loaded %d listfile entries for maps in %.1fs", len(files), time.Since(start).Seconds())
	}

	adtByDir := map[string]map[string]struct{}{}
	texByDir := map[string]map[string]struct{}{}
	localFileNameIndex := map[string]casc.ListfileEntry{}

	for _, entry := range files {
		fileName := strings.ToLower(strings.ReplaceAll(entry.FileName, `\`, `/`))
		if m := tileBlpRegex.FindStringSubmatch(fileName); m != nil {
			dir := m[1]
			x, _ := strconv.Atoi(m[2])
			y, _ := strconv.Atoi(m[3])
			if x >= 0 && x < 64 && y >= 0 && y < 64 {
				ensureSet(texByDir, dir)[tileKey(x, y)] = struct{}{}
				localFileNameIndex[fileName] = entry
			}
			continue
		}
		if m := tileAdtRegex.FindStringSubmatch(fileName); m != nil && strings.EqualFold(m[1], m[2]) {
			dir := strings.ToLower(m[1])
			x, _ := strconv.Atoi(m[3])
			y, _ := strconv.Atoi(m[4])
			if x >= 0 && x < 64 && y >= 0 && y < 64 {
				ensureSet(adtByDir, dir)[tileKey(x, y)] = struct{}{}
			}
		}
	}

	mapsMu.Lock()
	defer mapsMu.Unlock()
	mapsWithTiles = nil
	mapsByDir = map[string]*mapWithTiles{}
	fileNameIndex = localFileNameIndex

	for _, m := range baseMaps {
		dir := strings.ToLower(m.Dir)
		tilesMap := map[string]tileInfo{}

		for key := range adtByDir[dir] {
			x, y := parseTileKey(key)
			tilesMap[key] = tileInfo{X: x, Y: y, HasTexture: false}
		}
		for key := range texByDir[dir] {
			x, y := parseTileKey(key)
			if prev, ok := tilesMap[key]; ok {
				prev.HasTexture = true
				tilesMap[key] = prev
			} else {
				tilesMap[key] = tileInfo{X: x, Y: y, HasTexture: true}
			}
		}

		tiles := make([]tileInfo, 0, len(tilesMap))
		for _, t := range tilesMap {
			tiles = append(tiles, t)
		}
		withTiles := mapWithTiles{MapEntry: m, Tiles: tiles}
		mapsWithTiles = append(mapsWithTiles, withTiles)
		entry := withTiles
		mapsByDir[dir] = &entry
	}
	return nil
}

func ensureSet(m map[string]map[string]struct{}, dir string) map[string]struct{} {
	if m[dir] == nil {
		m[dir] = map[string]struct{}{}
	}
	return m[dir]
}

func tileKey(x, y int) string {
	return strconv.Itoa(x) + "," + strconv.Itoa(y)
}

func parseTileKey(key string) (int, int) {
	parts := strings.SplitN(key, ",", 2)
	x, _ := strconv.Atoi(parts[0])
	y, _ := strconv.Atoi(parts[1])
	return x, y
}

func ensureMapsIndex(ctx context.Context, d *Deps) error {
	mapsMu.RLock()
	if len(mapsWithTiles) > 0 {
		mapsMu.RUnlock()
		return nil
	}
	mapsMu.RUnlock()

	if err := d.Client.WaitUntilReady(ctx); err != nil {
		return err
	}

	mapsIndexMu.Lock()
	defer mapsIndexMu.Unlock()
	mapsMu.RLock()
	if len(mapsWithTiles) > 0 {
		mapsMu.RUnlock()
		return nil
	}
	mapsMu.RUnlock()
	return buildMapsIndex(ctx, d)
}

func resolveMap(key string) (id int, dir string, ok bool) {
	mapsMu.RLock()
	defer mapsMu.RUnlock()
	entry, found := mapsByDir[strings.ToLower(key)]
	if !found || entry == nil {
		return 0, "", false
	}
	return entry.ID, entry.Dir, true
}

func registerMaps(r Router, d *Deps) {
	go func() {
		defer d.startup.mapsFinished()
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Minute)
		defer cancel()
		if err := ensureMapsIndex(ctx, d); err != nil {
			log.Printf("maps index preload failed: %v", err)
		}
	}()

	registerMapGenerateRoutes(r, d)

	r.Get("/maps", func(w http.ResponseWriter, req *http.Request) {
		if err := ensureMapsIndex(req.Context(), d); err != nil {
			sendInternalError(w, err)
			return
		}
		mapsMu.RLock()
		buildKey := d.BuildKey(req.Context())
		etag := etagFromParts("maps", buildKey, strconv.Itoa(len(mapsWithTiles)))
		if matchNotModified(req, etag) {
			mapsMu.RUnlock()
			applyCascBuildCache(w, req, d.Config, buildKey, etag, false)
			writeNotModified(w, etag)
			return
		}
		out := make([]map[string]any, 0, len(mapsWithTiles))
		for _, m := range mapsWithTiles {
			out = append(out, map[string]any{
				"id": m.ID, "name": m.Name, "dir": m.Dir, "expansionID": m.ExpansionID,
			})
		}
		mapsMu.RUnlock()
		applyCascBuildCache(w, req, d.Config, buildKey, etag, false)
		sendJSON(w, http.StatusOK, out)
	})

	r.Get("/maps/{map}/wdt-mask", func(w http.ResponseWriter, req *http.Request) {
		if err := ensureMapsIndex(req.Context(), d); err != nil {
			sendInternalError(w, err)
			return
		}
		mapKey := strings.ToLower(chi.URLParam(req, "map"))
		mapsMu.RLock()
		entry := mapsByDir[mapKey]
		var tiles []tileInfo
		if entry != nil {
			tiles = entry.Tiles
		}
		buildKey := d.BuildKey(req.Context())
		etag := etagFromParts("wdt-mask", buildKey, mapKey, strconv.Itoa(len(tiles)))
		if matchNotModified(req, etag) {
			mapsMu.RUnlock()
			applyCascBuildCache(w, req, d.Config, buildKey, etag, false)
			writeNotModified(w, etag)
			return
		}
		mapsMu.RUnlock()
		applyCascBuildCache(w, req, d.Config, buildKey, etag, false)
		sendJSON(w, http.StatusOK, map[string]any{
			"map": chi.URLParam(req, "map"), "size": 64, "tiles": tiles,
		})
	})

	r.Post("/maps/{map}/creatures-check", func(w http.ResponseWriter, req *http.Request) {
		if err := assertDesktopOnly(d.Config.IsSharedHosting); err != nil {
			sendError(w, http.StatusForbidden, err.Error())
			return
		}
		mapID, _, ok := resolveMap(chi.URLParam(req, "map"))
		if !ok {
			sendError(w, http.StatusNotFound, "Unknown map")
			return
		}
		var body struct {
			Tiles []exportAdtTile `json:"tiles"`
		}
		if err := readJSONBody(req, &body); err != nil || len(body.Tiles) == 0 {
			sendError(w, http.StatusBadRequest, "Invalid request body")
			return
		}

		ordered := dedupeTiles(body.Tiles)
		if len(ordered) == 0 {
			sendError(w, http.StatusBadRequest, "No valid tiles provided")
			return
		}

		tileCoords := make([][2]int, len(ordered))
		checkedTiles := make([]exportAdtTile, len(ordered))
		for i, tile := range ordered {
			if tile.X < 0 || tile.X >= 64 || tile.Y < 0 || tile.Y >= 64 {
				sendError(w, http.StatusBadRequest, "Tile coordinates must be within 0..63")
				return
			}
			tileCoords[i] = [2]int{tile.X, tile.Y}
			checkedTiles[i] = tile
		}

		count, err := azerothcore.CountCreaturesInTiles(mapID, tileCoords)
		if err != nil {
			if strings.Contains(err.Error(), "azerothcore database not found") {
				sendJSON(w, http.StatusOK, map[string]any{
					"hasCreatures":  false,
					"creatureCount": 0,
					"checkedTiles":  checkedTiles,
				})
				return
			}
			sendInternalError(w, err)
			return
		}

		sendJSON(w, http.StatusOK, map[string]any{
			"hasCreatures":  count > 0,
			"creatureCount": count,
			"checkedTiles":  checkedTiles,
		})
	})

	r.Get("/maps/{map}/minimap/{x}/{y}", func(w http.ResponseWriter, req *http.Request) {
		mapDir := strings.ToLower(chi.URLParam(req, "map"))
		x, errX := strconv.Atoi(chi.URLParam(req, "x"))
		y, errY := strconv.Atoi(chi.URLParam(req, "y"))
		if errX != nil || errY != nil || x < 0 || x >= 64 || y < 0 || y >= 64 {
			sendError(w, http.StatusBadRequest, "x and y must be within 0..63")
			return
		}

		xs := pad2(x)
		ys := pad2(y)

		buildKey := ""
		if info, err := d.Client.GetCASCInfo(req.Context()); err == nil {
			buildKey = info.BuildKey
		}
		etag := md5Hex(buildKey + "|" + mapDir + "|" + strconv.Itoa(x) + "|" + strconv.Itoa(y))
		quotedETag := `"` + etag + `"`
		if matchNotModified(req, quotedETag) {
			applyCascBuildCache(w, req, d.Config, buildKey, quotedETag, true)
			writeNotModified(w, quotedETag)
			return
		}

		assetDir := d.ExportAssetDir()
		preexisting := MinimapPngPath(assetDir, buildKey, mapDir, xs, ys)
		if data, err := os.ReadFile(preexisting); err == nil {
			servePNG(w, req, d.Config, buildKey, data, quotedETag)
			return
		}

		blpPath := "world/minimaps/" + mapDir + "/map" + xs + "_" + ys + ".blp"
		mapsMu.RLock()
		file, found := fileNameIndex[blpPath]
		mapsMu.RUnlock()
		if !found || file.FileDataID == 0 {
			sendError(w, http.StatusNotFound, "Minimap tile not found")
			return
		}

		raw, err := d.Client.DownloadCascFile(req.Context(), file.FileDataID)
		if err != nil {
			sendInternalError(w, err)
			return
		}
		img, err := blp.NewBLPImage(buffer.From(raw))
		if err != nil {
			sendInternalError(w, err)
			return
		}
		pngBuf, err := img.ToPNG(0b1111, 0)
		if err != nil {
			sendInternalError(w, err)
			return
		}
		if err := os.MkdirAll(filepath.Dir(preexisting), 0o755); err == nil {
			_ = os.WriteFile(preexisting, pngBuf.Raw(), 0o644)
		}
		servePNG(w, req, d.Config, buildKey, pngBuf.Raw(), quotedETag)
	})
}

func pad2(n int) string {
	return fmt.Sprintf("%02d", n)
}

func md5Hex(s string) string {
	sum := md5.Sum([]byte(s))
	return hex.EncodeToString(sum[:])
}

func servePNG(w http.ResponseWriter, req *http.Request, cfg Config, activeBuild string, data []byte, etag string) {
	w.Header().Set("Content-Type", "image/png")
	applyCascBuildCache(w, req, cfg, activeBuild, etag, true)
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(data)
}
